Merge the NPointFunctions module into FlexibleSUSY

iolojz opened the pull request:
The NPointFunctions module can generate c++ code for scalar one-loop correlation functions in arbitrary FlexibleSUSY models. The interface is simple and the results have been subject to several tests.
As of now, there are two unit tests
  test/test_SM_npointfunctions.x
  test/test_MSSM_npointfunctions.x
that compare the calculated Higgs self-energies with the ones supplied by *_mass_eigenstates.

This module does, however, depend on FeynArts and FormCalc. Thus should perhaps be merged with some extra user flag (?). Furthermore it does not yet have enough assertions, so that incorrect usage might lead to a crash.

Future goals of the module are:
- add consistent assertions and documentation
- support fermion and vector legs
- remove sums over colour indices of internal fields - generate a colour factor instead

Note that the latter two goals are mathematically very similar but as of now we have not settled upon a definite method in approaching them.

Expander commented:
Dear Jobst,

for the moment FA/FC are not officially required to build FS. So, we definitely need a fall-back solution if FA/FC are not available. (At least as long as we haven't decided on whether FA/FC should be a dependency.)

If I understand your changes correctly, FA/FC are currently only required for the two new tests (SM and MSSM). Could you please add a fall-back mechanism so that these two tests are somehow "ignored" or "allways pass" if FA/FC are not available?  (I don't want the test suite to always fail because of a missing dependency that is currently optional.)

Many thanks in advance,
Alex

P.S.: If we decide that FA/FC are required to build FS, I think it would be most convenient to have an install script, which downloads these packages and installs them (a local folder inside the FS directory may be sufficient). This install script could be similar (or better :) ) to the `install-sarah.sh` script, which we currently ship to automatically download and install SARAH.

iolojz commented:
Yes. As I see it, we have two options if FA/FC are not there:
- we do not run the tests
- we ship FS with pregenerated code for the SM and the MSSM

The first solution would require some macros in test/module.mk, e.g
  `$(HAVE_FEYNARTS)` and `$(HAVE_FORMCALC)`
and then simply not add the tests when either is false.
The second solution is easy but then we hardcode model-specific names into the pregenerated code which might be bad if for some reason the SM and MSSM model files get changed.

What would you favour? How would a user specify (or do we detect) FA/FC?

Expander commented:
For the moment I'd prefer option 1 (we don't run the tests if FA/FC is not available).

You are right, we then need two variables in our `Makefile` which indicate whether FA/FC is available. Since this information is not model-specific, I'd propose to put this information into the `Makefile` in the FS root directory. Since this `Makefile` is generated by the `configure` script, the way to go would be to let the `configure` script check for FA/FC and fill corresponding tags in the `Makefile`.

It could work like this:

1) One adds these variables into `config/Makefile.in`, like so:

       ENABLE_FEYNARTS := @ENABLE_FEYNARTS@
       ENABLE_FORMCALC := @ENABLE_FORMCALC@

2) In the `configure` script run the `math` kernel (`$MATH`) and try to load FA/FC.

3) If FA/FC could be loaded, replace the markers from 1. by appropriate values (`yes`/`no`) when creating the root `Makefile`.

Expander commented on meta/NPointFunctions/createFAModelFile.m:
```diff
@@ -0,0 +1,48 @@
+
+CreateFAModelFile::usage="";
+
+Begin["Private`"];
```
I think this should read ``Begin["`Private`"]``

iolojz replied:
You are absolutely right! Fixed in commit b702a76066f65e9792ab83613ea55e397c471b6e.

iolojz commented:
Alright! We now have checks for FA/FC as well as a simple test to determine whether FC actually works (This is probably a good idea because it needs to be compiled by the user).
You mentioned, you wanted an install script for FA/FC. I do not know how to add such a thing or reference to FS. How would you want it to work in detail?

Expander commented:
> You mentioned, you wanted an install script for FA/FC. I do not know how to add such a thing or reference to FS. How would you want it to work in detail?

I'd propose we use the one provided at http://www.feynarts.de and document in the FS installation description how to use it. Maybe like this:

    wget http://www.feynarts.de/FeynInstall
    chmod 755 FeynInstall
    ./FeynInstall

iolojz commented:
Okay, so just documenting it in the README?

Expander commented:
Yes, that would be good. Maybe also in `doc/mainpage.dox.in` and `doc/install.dox`.

Please note, that the README file has been renamed in the development branch.

iolojz commented:
The above commit should do it. It seems, I do not have the proper permissions to resolve the file conflict though.

Expander commented:
I have remove `doc/install.dox`, because it was redundant with the description in the `README` file. Please put the description into the new `README.rst` file into the section `Building a FlexibleSUSY model`.

Expander commented:
Could you please also print the values of `ENABLE_FORMCALC` and `ENABLE_FEYNARTS` when running `make showbuild`?  This helps debugging.

Expander commented:
I've discovered one other little issue:  I run this:

    ./configure --with-models=SM --with-optional-modules=test
    make

and the code finishes compiling successfully. So far, so good.

But when I then run again

    make

the test `test/test_SM_npointfunctions.meta` is executed again:

~~~
make --debug
GNU Make 4.1
Built for x86_64-pc-linux-gnu
Copyright (C) 1988-2014 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Reading makefiles...
"math" -run "AppendTo[\$Path, \"./meta/\"]; Get[\"test/test_SM_npointfunctions.meta\"]; Quit[0];"
Mathematica 9.0 for Linux x86 (64-bit)
Copyright 1988-2012 Wolfram Research, Inc.
Comparing self-energy expressions generated by NPointFunctions to those from\
 
>   FlexibleSUSY ... 
SARAH 4.14.0
by Florian Staub, 2018
contributions by M. Gabelmann, M. D. Goodsell, K. Nickel

References:
  Comput.Phys.Commun.181 (2010) 1077-1086. (arXiv:0909.2863[hep-ph])
  Comput.Phys.Commun.182 (2011) 808-833. (arXiv:1002.0840[hep-ph])
  Comput.Phys.Commun.184 (2013) 1792-1809. (arXiv:1207.0906[hep-ph])
  Comput.Phys.Commun.185 (2014) 1773-1790. (arXiv:1309.7223[hep-ph])
Download and Documentation:
  http://sarah.hepforge.org

Start evaluation of a model with:
   Start["Name of Model"]
e.g. Start["MSSM"] or Start["NMSSM","CKM"]
To get a list with all installed models, use ShowModels

======================================================================
FlexibleSUSY 2.2.0
  by P. Athron, M. Bach, D. Harries, W. Kotlarski, T. Kwasnitza, 
  J.-h. Park, T. Steudtner, D. Stöckinger, A. Voigt, J. Ziebell
  2013-2018

References:
  Comput.Phys.Commun. 190 (2015) 139-172 (arxiv:1406.2319 [hep-ph])
  JHEP 1701 (2017) 079 (arXiv:1609.00371 [hep-ph])
  Comput.Phys.Commun. 230 (2018) 145-217 (arXiv:1710.03760 [hep-ph])

Download and Documentation:
  https://flexiblesusy.hepforge.org
======================================================================

Preparing arrays
 ... checking Directory: /home/avoigt/research/FS-jobst/sarah
 ... checking Directory: /home/avoigt/mathematica/SARAH-4.14.0/Models/
Model files loaded   
  Model    : SM
  Author(s): F.Staub
  Date     : 2014-11-06
******************************************************* 
Loading Susyno functions for the handling of Lie Groups 
Based on Susyno v.2.0 by Renato Fonseca (1106.5016)
webpage: web.ist.utl.pt/renato.fonseca/susyno.html
******************************************************* 
Initialization
Checking model files: Dynamic[DynamicCheckModelFiles]
Initialize gauge groups: Dynamic[DynamicInitGaugeG]
Initialize field: Dynamic[DynamicInitFields]
Preprocessing necessary information: Dynamic[DynamicInitMisc]
Checking for anomalies: Dynamic[DynamicCheckAnomalies]
Derive Lagrangian
Calculate kinetic Terms
  ... for scalars: Dynamic[DynamicKineticScalarNr]/6 (
 
>   Dynamic[DynamicKineticScalarName])
  ... for fermions: Dynamic[DynamicKineticFermionNr]/6 (
 
>   Dynamic[DynamicKineticFermionName])
Calculate vector self-interactions: Dynamic[DynamicVectorNr]/3 (
 
>   Dynamic[DynamicVectorName])
Calculate gauge transformations: Dynamic[DynamicGaugeTNr]/9 (
 
>   Dynamic[DynamicGaugeTName])
Eleminate sums
Calc Complete Lagrangian

Evolve States: GaugeES
Adding terms to the Lagrangian: 
 ... adding: -(Yu u . q . H) - Yd conj[H] . d . q - Ye conj[H] . e . l (
 
>   Dynamic[DynamicStatusAddTerms[-(Yu u . q . H) - Yd conj[H] . d . q - 
 
>      Ye conj[H] . e . l]])
                                  λ conj[H] . H . conj[H] . H
 ... adding: -(mu2 conj[H] . H) - ---------------------------- (
                                               2
 
>   Dynamic[DynamicStatusAddTerms[-(mu2 conj[H] . H) - 
 
       λ conj[H] . H . conj[H] . H
>      ----------------------------]])
                    2
Rotate Lagrangian: Dynamic[DynamicRotateLag[1]]/14
Derive ghost terms: 
  ... generate gauge fixing terms: Dynamic[DynamicGFnr[GaugeES]]/3 (
 
>   Dynamic[DynamicGFname[GaugeES]])
  ... calculate Ghost interactions
Calc Mixings of Matter Fields
Save information (Dynamic[DynamicSaveInfo[GaugeES]])

Evolve States: EWSB
Parametrize Higgs Sector
   Update gauge transformations: Dynamic[DynamicUGT[UGTvev[2]]]/17 (
 
>   Dynamic[DynamicUGTname[UGTvev[2]]])
Calc mass matrices gauge sector: Dynamic[DynamicMMgaugeNr[EWSB]]/2(
 
>   Dynamic[DynamicMMgaugeName[EWSB]])
   Update gauge transformations: Dynamic[DynamicUGT[UGTgaugeMM[2]]]/24 (
 
>   Dynamic[DynamicUGTname[UGTgaugeMM[2]]])
Rotate Lagrangian: Dynamic[DynamicRotateLag[2]]/14
Derive ghost terms: 
  ... generate gauge fixing terms: Dynamic[DynamicGFnr[EWSB]]/4 (
 
>   Dynamic[DynamicGFname[EWSB]])
  ... calculate Ghost interactions
Calc Mixings of Matter Fields
   Calculate mass matrices Dynamic[DynamicNrMass[{{{dL}, {conj[dR]}}, 
 
>      {{uL}, {conj[uR]}}, {{eL}, {conj[eR]}}}]]/3 (
 
>   Dynamic[DynamicNameMass[{{{dL}, {conj[dR]}}, {{uL}, {conj[uR]}}, 
 
>      {{eL}, {conj[eR]}}}]])
   Update gauge transformations: Dynamic[DynamicUGT[UGTmatterMM[2]]]/26 (
 
>   Dynamic[DynamicUGTname[UGTmatterMM[2]]])
Calculate Tadpole Equations
Save information (Dynamic[DynamicSaveInfo[EWSB]])

Finishing
Calculate final Lagrangian
Cleaning up
   ... add matrix products
   ... checking for massless particles
   ... calculating tree level masses (Dynamic[DynamicCalcTreeMasses])
   ... simplify mass matrices
Numerical calculations (if necessary)
   ... checking for spectrum file: Dynamic[DynamicSpectrumFileInput]
   ... reading parameter values and dependences
   ... calculate mixing matrices
Checking for CP even and odd scalars

All Done. SM is ready!
(Model initialized in 2.596s)


Are you unfamiliar with SARAH? Use SARAH`FirstSteps

----------------------------------------------------------------------
Loading FlexibleSUSY model file
----------------------------------------------------------------------
Found up-to-date RGE files.
Calculate non-supersymmetric RGEs
Read Beta Functions
Writing Mathematica code to evaluate RGEs

Finished with the calculation of the RGEs. Time needed: 0.096s
The results are saved in /home/avoigt/research/FS-jobst/Output/SM/RGEs/
Mass matrix for Hp does not exist!
Mass matrix for Fv does not exist!
Mass matrix for Ah does not exist!
Mass matrix for hh does not exist!
NPointFunctions`NPointFunction[]: Option ZeroExternalMomenta must be either\
 
>   True or False
Updating goal targets....
 File 'all' does not exist.
   File 'allsrc' does not exist.
  Must remake target 'allsrc'.
  Successfully remade target file 'allsrc'.
   File 'alllib' does not exist.
         Prerequisite 'run-metacode-SM' of target 'models/SM/SM_a_muon.hpp' does not exist.
        Must remake target 'models/SM/SM_a_muon.hpp'.
        Successfully remade target file 'models/SM/SM_a_muon.hpp'.
         Prerequisite 'run-metacode-SM' of target 'models/SM/SM_mass_eigenstates.hpp' does not exist.
        Must remake target 'models/SM/SM_mass_eigenstates.hpp'.
        Successfully remade target file 'models/SM/SM_mass_eigenstates.hpp'.
         Prerequisite 'run-metacode-SM' of target 'models/SM/cxx_qft/SM_qft.hpp' does not exist.
        Must remake target 'models/SM/cxx_qft/SM_qft.hpp'.
        Successfully remade target file 'models/SM/cxx_qft/SM_qft.hpp'.
         Prerequisite 'run-metacode-SM' of target 'models/SM/SM_info.hpp' does not exist.
        Must remake target 'models/SM/SM_info.hpp'.
        Successfully remade target file 'models/SM/SM_info.hpp'.
         Prerequisite 'run-metacode-SM' of target 'models/SM/SM_physical.hpp' does not exist.
        Must remake target 'models/SM/SM_physical.hpp'.
        Successfully remade target file 'models/SM/SM_physical.hpp'.
         Prerequisite 'run-metacode-SM' of target 'models/SM/SM_soft_parameters.hpp' does not exist.
        Must remake target 'models/SM/SM_soft_parameters.hpp'.
        Successfully remade target file 'models/SM/SM_soft_parameters.hpp'.
         Prerequisite 'run-metacode-SM' of target 'models/SM/SM_susy_parameters.hpp' does not exist.
        Must remake target 'models/SM/SM_susy_parameters.hpp'.
        Successfully remade target file 'models/SM/SM_susy_parameters.hpp'.
         Prerequisite 'run-metacode-SM' of target 'models/SM/SM_input_parameters.hpp' does not exist.
        Must remake target 'models/SM/SM_input_parameters.hpp'.
        Successfully remade target file 'models/SM/SM_input_parameters.hpp'.
         Prerequisite 'run-metacode-SM' of target 'models/SM/cxx_qft/SM_context_base.hpp' does not exist.
        Must remake target 'models/SM/cxx_qft/SM_context_base.hpp'.
        Successfully remade target file 'models/SM/cxx_qft/SM_context_base.hpp'.
         Prerequisite 'run-metacode-SM' of target 'models/SM/cxx_qft/SM_fields.hpp' does not exist.
        Must remake target 'models/SM/cxx_qft/SM_fields.hpp'.
        Successfully remade target file 'models/SM/cxx_qft/SM_fields.hpp'.
         Prerequisite 'run-metacode-SM' of target 'models/SM/cxx_qft/SM_vertices.hpp' does not exist.
        Must remake target 'models/SM/cxx_qft/SM_vertices.hpp'.
        Successfully remade target file 'models/SM/cxx_qft/SM_vertices.hpp'.
         Prerequisite 'run-metacode-SM' of target 'models/SM/SM_edm.hpp' does not exist.
        Must remake target 'models/SM/SM_edm.hpp'.
        Successfully remade target file 'models/SM/SM_edm.hpp'.
         Prerequisite 'run-metacode-SM' of target 'models/SM/SM_effective_couplings.hpp' does not exist.
        Must remake target 'models/SM/SM_effective_couplings.hpp'.
        Successfully remade target file 'models/SM/SM_effective_couplings.hpp'.
         Prerequisite 'run-metacode-SM' of target 'models/SM/SM_ewsb_solver_interface.hpp' does not exist.
        Must remake target 'models/SM/SM_ewsb_solver_interface.hpp'.
        Successfully remade target file 'models/SM/SM_ewsb_solver_interface.hpp'.
         Prerequisite 'run-metacode-SM' of target 'models/SM/SM_two_scale_ewsb_solver.hpp' does not exist.
        Must remake target 'models/SM/SM_two_scale_ewsb_solver.hpp'.
        Successfully remade target file 'models/SM/SM_two_scale_ewsb_solver.hpp'.
         Prerequisite 'run-metacode-SM' of target 'models/SM/SM_ewsb_solver.hpp' does not exist.
        Must remake target 'models/SM/SM_ewsb_solver.hpp'.
        Successfully remade target file 'models/SM/SM_ewsb_solver.hpp'.
         Prerequisite 'run-metacode-SM' of target 'models/SM/SM_observables.hpp' does not exist.
        Must remake target 'models/SM/SM_observables.hpp'.
        Successfully remade target file 'models/SM/SM_observables.hpp'.
         Prerequisite 'run-metacode-SM' of target 'models/SM/SM_slha_io.hpp' does not exist.
        Must remake target 'models/SM/SM_slha_io.hpp'.
        Successfully remade target file 'models/SM/SM_slha_io.hpp'.
         Prerequisite 'run-metacode-SM' of target 'models/SM/SM_model_slha.hpp' does not exist.
        Must remake target 'models/SM/SM_model_slha.hpp'.
        Successfully remade target file 'models/SM/SM_model_slha.hpp'.
         Prerequisite 'run-metacode-SM' of target 'models/SM/SM_two_scale_convergence_tester.hpp' does not exist.
        Must remake target 'models/SM/SM_two_scale_convergence_tester.hpp'.
        Successfully remade target file 'models/SM/SM_two_scale_convergence_tester.hpp'.
         Prerequisite 'run-metacode-SM' of target 'models/SM/SM_convergence_tester.hpp' does not exist.
        Must remake target 'models/SM/SM_convergence_tester.hpp'.
        Successfully remade target file 'models/SM/SM_convergence_tester.hpp'.
         Prerequisite 'run-metacode-SM' of target 'models/SM/SM_two_scale_model.hpp' does not exist.
        Must remake target 'models/SM/SM_two_scale_model.hpp'.
        Successfully remade target file 'models/SM/SM_two_scale_model.hpp'.
         Prerequisite 'run-metacode-SM' of target 'models/SM/SM_model.hpp' does not exist.
        Must remake target 'models/SM/SM_model.hpp'.
        Successfully remade target file 'models/SM/SM_model.hpp'.
         Prerequisite 'run-metacode-SM' of target 'models/SM/SM_two_scale_high_scale_constraint.hpp' does not exist.
        Must remake target 'models/SM/SM_two_scale_high_scale_constraint.hpp'.
        Successfully remade target file 'models/SM/SM_two_scale_high_scale_constraint.hpp'.
         Prerequisite 'run-metacode-SM' of target 'models/SM/SM_high_scale_constraint.hpp' does not exist.
        Must remake target 'models/SM/SM_high_scale_constraint.hpp'.
        Successfully remade target file 'models/SM/SM_high_scale_constraint.hpp'.
         Prerequisite 'run-metacode-SM' of target 'models/SM/SM_two_scale_initial_guesser.hpp' does not exist.
        Must remake target 'models/SM/SM_two_scale_initial_guesser.hpp'.
        Successfully remade target file 'models/SM/SM_two_scale_initial_guesser.hpp'.
         Prerequisite 'run-metacode-SM' of target 'models/SM/SM_initial_guesser.hpp' does not exist.
        Must remake target 'models/SM/SM_initial_guesser.hpp'.
        Successfully remade target file 'models/SM/SM_initial_guesser.hpp'.
         Prerequisite 'run-metacode-SM' of target 'models/SM/SM_two_scale_low_scale_constraint.hpp' does not exist.
        Must remake target 'models/SM/SM_two_scale_low_scale_constraint.hpp'.
        Successfully remade target file 'models/SM/SM_two_scale_low_scale_constraint.hpp'.
         Prerequisite 'run-metacode-SM' of target 'models/SM/SM_two_scale_susy_scale_constraint.hpp' does not exist.
        Must remake target 'models/SM/SM_two_scale_susy_scale_constraint.hpp'.
        Successfully remade target file 'models/SM/SM_two_scale_susy_scale_constraint.hpp'.
         Prerequisite 'run-metacode-SM' of target 'models/SM/SM_low_scale_constraint.hpp' does not exist.
        Must remake target 'models/SM/SM_low_scale_constraint.hpp'.
        Successfully remade target file 'models/SM/SM_low_scale_constraint.hpp'.
         Prerequisite 'run-metacode-SM' of target 'models/SM/SM_susy_scale_constraint.hpp' does not exist.
        Must remake target 'models/SM/SM_susy_scale_constraint.hpp'.
        Successfully remade target file 'models/SM/SM_susy_scale_constraint.hpp'.
         Prerequisite 'run-metacode-SM' of target 'models/SM/SM_weinberg_angle.hpp' does not exist.
        Must remake target 'models/SM/SM_weinberg_angle.hpp'.
        Successfully remade target file 'models/SM/SM_weinberg_angle.hpp'.
         Prerequisite 'run-metacode-SM' of target 'models/SM/SM_two_scale_spectrum_generator.hpp' does not exist.
        Must remake target 'models/SM/SM_two_scale_spectrum_generator.hpp'.
        Successfully remade target file 'models/SM/SM_two_scale_spectrum_generator.hpp'.
         Prerequisite 'run-metacode-SM' of target 'models/SM/SM_spectrum_generator.hpp' does not exist.
        Must remake target 'models/SM/SM_spectrum_generator.hpp'.
        Successfully remade target file 'models/SM/SM_spectrum_generator.hpp'.
         Prerequisite 'run-metacode-SM' of target 'models/SM/SM_spectrum_generator_interface.hpp' does not exist.
        Must remake target 'models/SM/SM_spectrum_generator_interface.hpp'.
        Successfully remade target file 'models/SM/SM_spectrum_generator_interface.hpp'.
         Prerequisite 'run-metacode-SM' of target 'models/SM/SM_utilities.hpp' does not exist.
        Must remake target 'models/SM/SM_utilities.hpp'.
        Successfully remade target file 'models/SM/SM_utilities.hpp'.
  Must remake target 'alllib'.
  Successfully remade target file 'alllib'.
   File 'allexec' does not exist.
  Must remake target 'allexec'.
  Successfully remade target file 'allexec'.
   File 'allll' does not exist.
  Must remake target 'allll'.
  Successfully remade target file 'allll'.
Must remake target 'all'.
Successfully remade target file 'all'.
~~~

So, it seems that there is still some dependency issue, which forces `make` to re-run the test. Besides this, there is this warning/error message:

    NPointFunctions`NPointFunction[]: Option ZeroExternalMomenta must be either True or False

iolojz commented:
Unfortunately, I am unable to reproduce the behaviour you mention. On my machine, it only executes the .meta code once and without errors.  So if this is a real issue, it probably has to do with some older Mathematica versions (just guessing...). Could you please try the same things again after a `git clean -xfd`?

Expander commented:
> Unfortunately, I am unable to reproduce the behaviour you mention. On my machine, it only executes the .meta code once and without errors. So if this is a real issue, it probably has to do with some older Mathematica versions (just guessing...). Could you please try the same things again after a `git clean -xfd`?

I've investigated a little further and the issue is, that the command

    "math" -run "AppendTo[\$Path, \"./meta/\"]; Get[\"test/test_SM_npointfunctions.meta\"]; Quit[0];"

fails with the error message:

    NPointFunctions`NPointFunction[]: Option ZeroExternalMomenta must be either True or False

Afther this error message the script aborts and the file `test/test_SM_npointfunctions.cpp` is not created. When `make` is run again, it tries to generate the file again, but Mathematica fails again with the same error message.

The other issue here is, that the script aborts with exit code 0, which signals that everything went fine. This is why I didn't spot this failure when running the test suite.

iolojz commented:
I fixed the merge conflicts and put the FA/FC documentation in the README.rst file. I believe it fits better into the 'Building FlexibleSUSY' section though - below 'Installation of SARAH'.

I continue to be unable to reproduce your issues though... When I run
    `git clean -xfd`
    `./createmodel --name=SM`
    `./configure --with-models=SM --with-optional-modules=test`
    `make`
everything finishes without error messages. If I then attempt to run `make` again nothing happens - as expected. Maybe you somehow have a polluted cache where something in Output/SM is corrupt. Removing Output/SM/EWSB/FeynArts, Output/SM/EWSB/NPointFunctions and test/test_SM_npointfunctions.cpp will likely fix your issue.

Expander commented:
Many thanks for looking into this!

The issue is not that the source tree is polluted.  The problem is, that the `math` command fails with the error message

    NPointFunctions`NPointFunction[]: Option ZeroExternalMomenta must be either True or False

After this error the `math` command aborts with the exit code 0 and the file `test/test_SM_npointfunctions.cpp` is not created. So, to fix this we should understand why the script fails with the above error message. :)

iolojz commented:
Well, I try to do what I can... But on my machine I do not get any error messages. Also `test/test_SM_npointfunctions.cpp` is created as it is supposed to be.

Expander commented:
I see.  :)  So, it may have to do with my setup.  Do you have a machine with Mathematica 9 to test this on?

iolojz commented:
Well... Apparently AllTrue[] is also a Mathematica 10+ function... I will
fix it when I get back to my computer.

Expander commented:
Now the code continues and runs further. It now gives errors:

~~~
StringSplit::strse: 
   String or list of strings expected at position 1 in 
    StringSplit[NPointFunctions`Private`ReadString[InputStream[
       /home/avoigt/research/FS-jobst/Output/SM/EWSB/FeynArts/ParticleNamesFey
        nArts.dat, 314]], ].


StringMatchQ::strse: 
   String or list of strings expected at position 1 in 
    StringMatchQ[NPointFunctions`Private`ReadString[InputStream[
       /home/avoigt/research/FS-jobst/Output/SM/EWSB/FeynArts/ParticleNamesFey
        nArts.dat, 314]], ___~~hh:~~___].

StringSplit::argb: 
   StringSplit called with 0 arguments; between 1 and 3
     arguments are expected.
~~~

iolojz commented:
I begin to believe you (or Mathematica) is doing this to me on purpose ^^
Obviously your version of Mathematica knows nothing about ReadString[] which was introduced in version 10.0 it seems... I will try to work around that. Usually I only need individual lines anyway which could be provided by Read[] which is available since Mathematica's inception.

Expander commented:
Some time ago I was thinking whether we should add a compatibility module, which defines certain Mathematica functions in case the user has an older version. Maybe like this (pseudo code):
~~~
If[MathematicaVersion < 10,
   ReadString[args__] := Read[...]
  ]
~~~

Expander commented:
I think you're right, using the old interfaces would be simpler. But that means we need to test on old Mathematica versions. :)

I pulled your recent changes and ran the code again. I'm still getting some errors from kernel 2:

~~~
From KernelObject[2, local]:
FeynArts 3.9

From KernelObject[2, local]:
by Hagen Eck, Sepp Kueblbeck, and Thomas Hahn

From KernelObject[2, local]:
last revised 4 Apr 15

From KernelObject[2, local]:


From KernelObject[2, local]:
FormCalc 9.7 (4 Dec 2018)

From KernelObject[2, local]:
by Thomas Hahn

From KernelObject[2, local]:


From KernelObject[2, local]:
loading generic model file\
 
>   /home/avoigt/mathematica/FeynArts/Models/Lorentz.g\
 
      en

From KernelObject[2, local]:
> $GenericMixing is OFF

From KernelObject[2, local]:
generic model {Lorentz} initialized

From KernelObject[2, local]:


From KernelObject[2, local]:
loading classes model file\
 
>   /home/avoigt/research/FS-jobst/Output/SM/EWSB/Feyn\
 
      Arts/SMEWSB.mod

From KernelObject[2, local]:
> 43 particles (incl. antiparticles) in 16 classes

From KernelObject[2, local]:
> $CounterTerms are ON

From KernelObject[2, local]:
> 94 vertices

From KernelObject[2, local]:
classes model {/home/avoigt/research/FS-jobst/Output/SM/EWSB/FeynArts/SMEWSB}
 
      initialized

From KernelObject[2, local]:


From KernelObject[2, local]:
inserting at level(s) {Generic, Classes}

From KernelObject[2, local]:
> Top. 1: 2 Generic, 5 Classes insertions

From KernelObject[2, local]:
> Top. 2: 5 Generic, 14 Classes insertions

From KernelObject[2, local]:
in total: 7 Generic, 19 Classes insertions

From KernelObject[2, local]:


From KernelObject[2, local]:
creating amplitudes at level(s) {Generic, Classes}

From KernelObject[2, local]:
> Top. 1: 2 Generic, 5 Classes amplitudes

From KernelObject[2, local]:
> Top. 2: 5 Generic, 14 Classes amplitudes

From KernelObject[2, local]:
in total: 7 Generic, 19 Classes amplitudes

From KernelObject[2, local]:
Except::named: Named pattern variables are not allowed in the first argument
     of Except[s_Symbol /; Context[s] == System`].
From KernelObject[2, local]:

preparing FORM code in\
 
     /home/avoigt/research/FS-jobst/Output/SM/EWSB/FormCalc/fc-amp-2001.frm

From KernelObject[2, local]:
running FORM... 

From KernelObject[2, local]:
ReadForm::formerror: Unmatched quotes for preprocessor variable mass,

Part::partd: Part specification $Aborted[[1]] is longer than depth of object.
Warning: Using FlexibleSUSY loop functions will only remap A0, B0, C0 and D0.
Warning: FlexibleSUSY loop functions C0 and D0 require zero external momenta.

Part::partd: Part specification $Aborted[[1,1]]
     is longer than depth of object.

Part::partd: Part specification $Aborted[[1,2]]
     is longer than depth of object.

General::stop: Further output of Part::partd
     will be suppressed during this calculation.

Part::pspec: Part specification $Aborted
     is neither a machine-sized integer nor a list of machine-sized integers.

Part::pspec: Part specification $Aborted
     is neither a machine-sized integer nor a list of machine-sized integers.

Part::pspec: Part specification $Aborted
     is neither a machine-sized integer nor a list of machine-sized integers.

General::stop: Further output of Part::pspec
     will be suppressed during this calculation.

StringJoin::string: 
   String expected at position 1 in StringJoin[$Aborted[[1,1,$Aborted,1,2]]].

StringJoin::string: 
   String expected at position 2 in nPoint<>$Aborted[[1,1,$Aborted,1,2]].

StringJoin::string: 
   String expected at position 2 in 
    nPoint<>$Aborted[[1,1,$Aborted,1,2]]<> helper{ model, indices, momenta };.
                                          return helper.calculate();

General::stop: Further output of StringJoin::string
     will be suppressed during this calculation.

StringReplace::strse: 
   String or list of strings expected at position 1 in 
    StringReplace[ReadString[InputStream[/home/avoigt/research/FS-jobst/test/t
        est_SM_npointfunctions.cpp.in, 279]], 
     {@NPointFunctions_Selfenergy_hh_1loop@ -> <<12>><><<8>>, <<3>>}].
~~~

The first error is:

~~~
Except::named: Named pattern variables are not allowed in the first argument
     of Except[s_Symbol /; Context[s] == System`].
~~~

iolojz commented:
We would need to test on older versions anyway to make sure our compatibility functions work properly. SO I do not see a benefit of doing so. ALso, in the documentation one can usually find out which version introduced a feature.  

Anyway... Wow, that problem looks interesting! Especially because I seriously have no clue at all how you managed to get that error message... In my code, there is not a single Except[] call, nowhere a pattern 's_Symbol' and nowhere a string "System`". In fact I grepped over the whole source tree and found pretty much nothing that would even remotely fit. Are you certain, it is my code that you are executing :D

Expander commented:
> Are you certain, it is my code that you are executing :D

Could it be that this error comes from inside FA/FC or even SARAH?

Expander commented:
It seems to appear in FormCalc:

~~~
$ rg "Except\[s_Symbol"
FormCalc-9.7/FormCalc.m
4093:  Cases[{r}, Except[s_Symbol /; Context[s] == "System`"],
~~~

I think we should write Thomas Hahn to fix this compatibility problem.

iolojz commented:
Lovely! (Merry Christmas btw) :D
Yes, we need to do so.

Am Mo., 24. Dez. 2018, 07:32 hat Alexander Voigt <notifications@github.com>
geschrieben:

> It seems to appear in FormCalc:
>
> $ rg "Except\[s_Symbol"
> FormCalc-9.7/FormCalc.m
> 4093:  Cases[{r}, Except[s_Symbol /; Context[s] == "System`"],
>
> I think we should write Thomas Hahn to fix this compatibility problem.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/FlexibleSUSY/FlexibleSUSY/pull/37#issuecomment-449692494>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMJ_ZG_I3uQF83N15kOlk1LJ3mUn1F-Hks5u8HUQgaJpZM4ZDIpw>
> .
>

iolojz commented:
But this is then probably an issue with version 9.7 and maybe with 9.6.
IIRC I used 9.5 with Mathematica 8 at some point... If this is true, we are
lucky and can just blacklist the version(s) in question with an error
message.

I will look into it after Christmas, but as I do not have Mathematica 9 I
do not know if I can trigger it. OTOH you already have found the line of
code where it occurs so I guess we could just tell him that his code
requires a more recent Mathematica version - which might not be an issue
for him.

Am Mo., 24. Dez. 2018, 07:32 hat Alexander Voigt <notifications@github.com>
geschrieben:

> It seems to appear in FormCalc:
>
> $ rg "Except\[s_Symbol"
> FormCalc-9.7/FormCalc.m
> 4093:  Cases[{r}, Except[s_Symbol /; Context[s] == "System`"],
>
> I think we should write Thomas Hahn to fix this compatibility problem.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/FlexibleSUSY/FlexibleSUSY/pull/37#issuecomment-449692494>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMJ_ZG_I3uQF83N15kOlk1LJ3mUn1F-Hks5u8HUQgaJpZM4ZDIpw>
> .
>

Expander commented:
Merry Christmas!

I've now switched to FC 9.5 and this seems to run ok, at least it finishes with the message:

~~~
From KernelObject[2, local]:
running FORM... 

From KernelObject[2, local]:
ok
~~~

After that step, however, an error appears again, which has to do with the calls to `ReadString[]` in `test/test_SM_npointfunctions.meta`:

~~~
Read Vertex Lists

StringReplace::strse: 
   String or list of strings expected at position 1 in 
    StringReplace[ReadString[InputStream[/home/avoigt/research/FS-jobst/test/t
        est_SM_npointfunctions.cpp.in, 279]], 
     {@NPointFunctions_Selfenergy_hh_1loop@ -> clas<<8>>hhhh , <<3>>}].
~~~

I guess the fix would be to avoid calling `ReadString[]` also in the `test/*.meta` files.

iolojz commented:
Done!

Expander commented:
Hi Jobst,

I've run the SM and the MSSM with your last changes with Mathematica 9 and FormCalc 9.5 and it runs nicely, even the tests. Many thanks!

Before I merge this branch, I have a couple of final questions, just to make sure everything works as expected:

* In its current from FS runs and compiles (even the tests) even if FA/FC is not available, right?  I.e. for the moment we have full backwards compatibility.
* Your changes do not conflict with the developments of Wojciech's and Dylan's work, right?

iolojz commented:
Great! Yes, things seem to work as they are supposed to. Regarding FA/FC presence:
- When FA and/or FC is avaible through `Needs[]`, we attempt to read the version. If this fails the configure script aborts with an error message. Is this okay, or do you want it to simply disable support in such cases?
- When FA and FC are enabled, we perform a test calculation of the Higgs self-energy in the SM. If this fails: same question as above.

If either FA/FC are/get disabled, `test_SM_npointfunctions.cpp` and `test_MSSM_npointfunctions.cpp` are not added to `TEST_SRC`. Thus they are not added to the list of tests. In fact, the files do not even get created in this case. Since they are not listed as tests, they cannot fail.

Woijciech and Dylan use the `CXXDiagrams` module which is almost unchanged by the preceeding commits (no interface changes though) - actually one could factor out these commits as they do themselves not have to do with NPointFunctions... But, if you give the green light, we can just as well merge things directly.

Expander commented:
Dear Jobst,

many thanks for your explanations!

Concerning your two questions: Could you please disable the FA/FC support (for now), if `Needs[]` works, but the version does not match? In this way we have maximum backwards compatibility for now. (Though, we may later change that when we require FA/FC to generate diagrams for the observables.)

Since today also the nightly test server works again. My plan is to wait until tomorrow to see if the current development branch works. If it does (and you made the change from above), then I'll merge this pull request into development. I think that would be cleanest.

Best regards,
Alex

Expander commented:
One last thing: There seem to be merge conflicts. Would you mind having a look what's going on there? (Probably nothing crucial.)

iolojz commented:
I just realized, I was wrong with my first point from before, sorry... By default `enable_feynarts` and `enable_formcalc` are set to `automatic` so that in case of an unreadable version or too low version the support is non-fatally disabled. Only if the user sets them to `yes` such errors become critical.

Point two is now fixed: When `enable_formcalc` is set to `automatic` (the default) and the test calculation fails the FormCalc support is non-fatally disabled.

Furthermore, in case of `enable_feynarts=no` or `automatic` that gets resolved to `no`:
- if `enable_formcalc=automatic`, we disable FormCalc support because we cannot perform a test calculation without FeynArts
- if `enable_formcalc=yes`, we enable FormCalc support because the user explicitly asked for it, even though we cannot perform a test calculation without FeynArts

The conflicts are resolved. I am looking forward to the merge ^^

Expander commented:
Dear Jobst,

unfortunately I have bad news: The CMSSM does not compile on your branch.  I've run:

~~~
git clean -xfd
./createmodel --name=CMSSM -f
./configure --with-models=CMSSM --with-optional-modules=test
make -j4
~~~

and I get the following error message:

~~~
models/CMSSM/CMSSM_two_scale_high_scale_constraint.cpp: In member function ‘void flexiblesusy::CMSSM_high_scale_constraint<flexiblesusy::Two_scale>::update_scale()’:
models/CMSSM/CMSSM_two_scale_high_scale_constraint.cpp:475:34: error: ‘List’ was not declared in this scope
    scale = currentScale*Exp(List());
                                  ^
Makefile:236: recipe for target 'models/CMSSM/CMSSM_two_scale_high_scale_constraint.o' failed
~~~

Could it be that some of your changes to `FlexibleSUSY.m` broke this?

iolojz commented:
Yes, you are right. I had added a more explicit scoping to some global variables in the ```Private` ``` context in `meta/FlexibleSUSY.m`. I had referenced them with ```FlexibleSUSY` ``` which obviously was wrong. That should be fixed now.

I also removed the global private variable `allOutputParameters` because it only was used for the `DebugPrint[]` call, that I simply moved to the place of the local variable `allOutputParameters`.

Expander commented:
Many thanks, this fixes the CMSSM!
I will now run some more tests and later merge your branch if everything works as expected.
